extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/sql/LiveCatalogResolver.java

```diff
@@ -0,0 +1,211 @@
+      return dsMetadata == null ? null : new DatasourceTable(dsMetadata);
+    }
+    if (dsMetadata == null) {
+      // Datasource exists only in the catalog: no physical segments.
+      return emptyDatasource(name, dsSpec);
+    } else {
+      // Datasource exists as both segments and a catalog entry.
+      return mergeDatasource(dsMetadata, dsSpec);
+    }
+  }
+
+  private DruidTable emptyDatasource(String name, DatasourceFacade dsSpec)
+  {
+    RowSignature.Builder builder = RowSignature.builder();
+    Map<String, EffectiveColumnMetadata> columns = new HashMap<>();
+    boolean hasTime = false;
+    for (ColumnSpec col : dsSpec.columns()) {
+      EffectiveColumnMetadata colMetadata = columnFromCatalog(col, null);
+      if (colMetadata.name().equals(Columns.TIME_COLUMN)) {
+        hasTime = true;
+      }
+      builder.add(col.name(), colMetadata.druidType());
+      columns.put(col.name(), colMetadata);
+    }
+    if (!hasTime) {
+      columns.put(Columns.TIME_COLUMN, new EffectiveColumnMetadata(
+          Columns.TIME_COLUMN,
+          ColumnType.LONG
+      ));
+      builder = RowSignature.builder()
+          .add(Columns.TIME_COLUMN, ColumnType.LONG)
+          .addAll(builder.build());
+    }
+
+    final PhysicalDatasourceMetadata mergedMetadata = new PhysicalDatasourceMetadata(
+          new TableDataSource(name),
+          builder.build(),
+          false, // Cannot join to an empty table
+          false // Cannot broadcast an empty table
+    );
+    return new DatasourceTable(
+        mergedMetadata.getRowSignature(),
+        mergedMetadata,
+        new EffectiveMetadata(dsSpec, columns, true)
+    );
+  }
+
+  private EffectiveColumnMetadata columnFromCatalog(ColumnSpec col, ColumnType physicalType)
+  {
+    ColumnType type = Columns.druidType(col);
+    if (type != null) {
+      // Use the type that the user provided.
+    } else if (physicalType == null) {
+      // Corner case: the user has defined a column in the catalog, has
+      // not specified a type (meaning the user wants Druid to decide), but
+      // there is no data at this moment. Guess String as the type for the
+      // null values. If new segments appear between now and execution, we'll
+      // convert the values to string, which is always safe.
+      type = ColumnType.STRING;
+    } else {
+      type = physicalType;
+    }
+    return new EffectiveColumnMetadata(col.name(), type);
+  }
+
+  private DruidTable mergeDatasource(
+      final PhysicalDatasourceMetadata dsMetadata,
+      final DatasourceFacade dsSpec)
+  {
+    final RowSignature physicalSchema = dsMetadata.getRowSignature();
+    Set<String> physicalCols = new HashSet<>(physicalSchema.getColumnNames());
+
+    // Merge columns. All catalog-defined columns come first,
+    // in the order defined in the catalog.
+    final RowSignature.Builder builder = RowSignature.builder();
+    Map<String, EffectiveColumnMetadata> columns = new HashMap<>();
```

Comment:
Does the insertion of keys, values into hashMap preserve order that it was added? I assume that the order of columns  of the effectiveMetadata is important?